pkg/messaging/direct_messaging.go

```diff
@@ -233,27 +243,189 @@ func (d *directMessaging) invokeRemote(ctx context.Context, appID, appNamespace,
 	start := time.Now()
 	diag.DefaultMonitoring.ServiceInvocationRequestSent(appID, req.Message().Method)
 
-	var resp *internalv1pb.InternalInvokeResponse
-	defer func() {
-		if resp != nil {
-			diag.DefaultMonitoring.ServiceInvocationResponseReceived(appID, req.Message().Method, resp.Status.Code, start)
-		}
-	}()
+	var imr *invokev1.InvokeMethodResponse
+	if !d.isStreamingEnabled {
+		imr, err = d.invokeRemoteUnary(ctx, clientV1, req, opts)
+	} else {
+		imr, err = d.invokeRemoteStream(ctx, clientV1, req, appID, opts)
+	}
+
+	// Diagnostics
+	if imr != nil {
+		diag.DefaultMonitoring.ServiceInvocationResponseReceived(appID, req.Message().Method, imr.Status().Code, start)
+	}
 
+	return imr, teardown, err
+}
+
+func (d *directMessaging) invokeRemoteUnary(ctx context.Context, clientV1 internalv1pb.ServiceInvocationClient, req *invokev1.InvokeMethodRequest, opts []grpc.CallOption) (*invokev1.InvokeMethodResponse, error) {
 	pd, err := req.ProtoWithData()
 	if err != nil {
-		return nil, teardown, fmt.Errorf("failed to read data from request object: %w", err)
```

Comment:
Why remove `teardown` here?

Reply:
`invokeRemoteUnary` and `invokeRemoteStream` are both invoked by `invokeRemote`, which returns `teardown`. See: https://github.com/dapr/dapr/pull/5706/files/2aae9dabd9c916f0c8b177104f98ade516071af2#diff-e74cf36c876319f39e2be04a7514c9cff0058ce19cad497e3f41f3f182a3ad73R246-R258